spec/lucky/action_spec.cr

```diff
@@ -82,6 +82,15 @@ class PlainText::Index < Lucky::Action
   end
 end
 
+class OptionalParams::Index < Lucky::Action
+  optional_param page
+  optional_param with_default, default: "default"
```

Comment:
Sorry to tack this on so late, but I think this can be changed to make it more flexible and be a bit more like other parts of Lucky/Crystal. Here's what I'm thinking:

```crystal
param page : Int32 # A required param
param page : Int32? # Not required, defaults to `nil` and is not added to query params if `nil`
param page : Int32 = 1 # Not required because if it is empty, defaults to `1`
```

This will be a bit trickier to implement, but is a bit more intuitive and powerful. It will also allow only the correct types and will parse them for you so you don't have to manually call `to_i`.

### Here is how it works:

LuckyRecord extends a number of built in types so that they know how to parse strings. You can see that in action here: https://github.com/luckyframework/lucky_record/tree/master/src/lucky_record/charms and specifically: https://github.com/luckyframework/lucky_record/blob/bfd25d77a24bae98a321c50eb5601bd77ce299b7/src/lucky_record/charms/int32_extensions.cr#L10-L14

Since Lucky depends on LuckyRecord we can use those extensions to parse string params into types like `Bool`, `Int32`, `String` or `Time`. 

It is a bit tricky to figure out how to do this, so here is a rough example that should show how this might work:

```crystal
# https://crystal-lang.org/api/0.23.1/Crystal/Macros/TypeDeclaration.html
macro page(type_declaration)
  {% PARAM_DECLARATIONS << type_declaration %}
end

param page : Int32?
{% for type_declaration in PARAM_DECLARATIONS %}
  {{ type_declaration.var }} # would be `page`
  {{ type_declaration.var }} # would be a type union. See example: https://github.com/luckyframework/lucky_migrator/blob/be44188ec6da68a0de9e9fc0c9494f8fd8b899d1/src/lucky_migrator/alter_table_statement.cr#L56
  {{ type_declaration.default }} # would be `nil` in this case
  # Parse a param like this
  # Note: type.first is for Union types. If you did `param page : Int32` you would do just `type`
  {{ type_declaration.type.first }}::Lucky.parse!(params.get(:{{ type_declaration.var }}))
{% end %}
```

Feel free to ping me in Gitter or comment here if this is unclear.

Reply:
Oooh thats interesting. Yep I definetely like your proposed syntax. Moving this into the type system makes so much sense. Can you tell I'm more fluent in dynamic langs? :D :D I will hack on this a bit over the coming days, lets see what we can make of it! Thanks for all your pointers.